core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/MigrationUtils.scala

```diff
@@ -16,42 +19,59 @@ object MigrationUtils {
-    final def fromFields(outcome: Option[String], message: Option[String]): Either[String, Option[Outcome]] = {
-      type EitherStringT[T] = Either[String, T]
-      outcome.traverse[EitherStringT, Outcome] {
+    final def fromFields(outcome: Option[String], message: Option[String]): Either[String, Option[Outcome]] =
+      outcome.traverse[Either[String, *], Outcome] {
         case "Success" => Right(Success)
         case "Failure" => Right(Failure(message.getOrElse("")))
         case other => Left(s"""Failed to read outcome: unknown value -- "$other"""")
       }
-    }
 
     final def toTuple(outcome: Outcome): (Option[String], Option[String]) = outcome match {
       case Success => ("Success".some, None)
       case Failure(msg) => ("Failure".some, msg.some)
     }
   }
 
+  object Implicits {
+    implicit val semigroupOutcome: Semigroup[Outcome] = (a, b) => a match {
+        case Success => b
+        case Failure(msgA) => b match {
+          case Success => a
+          case Failure(msgB) => Failure (msgA ++ "\n" ++ msgB)
+        }
+      }
 
-  def unsafeFromEither[A, B](f: A => Either[String, B], a: A): B = f(a) match {
-    case Right(r) => r
-    case Left(msg) => throw new RawlsException(msg)
-  }
+    implicit class IgnoreResultExtensionMethod[+R, +S <: NoStream, -E <: Effect](action: DBIOAction[R, S, E]) {
+      /** Ignore the result of the DBIOAction and return unit */
+      def ignore: DBIOAction[Unit, NoStream, E with Effect] = action >> DBIOAction.successful()
+    }
 
 
-  implicit class IgnoreResultExtensionMethod[+R, +S <: NoStream, -E <: Effect](action: DBIOAction[R, S, E]) {
-    /** Ignore the result of the DBIOAction and return unit */
-    def ignore: DBIOAction[Unit, NoStream, E with Effect] = action >> DBIOAction.successful()
-  }
+    import slick.jdbc.MySQLProfile.api._
+    implicit class InsertExtensionMethod[E, A, C[_]](query: Query[E, A, C]) {
+      /** alias for `+=` supporting dot syntax */
+      def insert(value: A) = query += value
+    }
 
 
-  import slick.jdbc.MySQLProfile.api._
-  implicit class InsertExtensionMethod[E, T, C[_]](query: Query[E, T, C]) {
-    /** alias for `+=` supporting dot syntax */
-    def insert(value: T) = query += value
+    implicit class FutureToIO[+A](future: => Future[A]) {
+      def io: IO[A] = IO.fromFuture(IO(future))
+    }
   }
 
 
-  implicit class FutureToIO[+T](future: => Future[T]) {
-    def io: IO[T] = IO.fromFuture(IO(future))
+  def unsafeFromEither[A, B](f: A => Either[String, B], a: A): B = f(a) match {
+    case Right(r) => r
+    case Left(msg) => throw new RawlsException(msg)
   }
+
+
+  final case class WorkspaceMigrationException(message: String,
+                                               data: Map[String, Any] = Map.empty,
+                                               cause: Throwable = null)
+    extends RawlsException(
+      message = JsObject((Map("message" -> message) ++ data).mapValues(s => new JsString(s.toString))).toString,
```

Comment:
I'm converting to JSON because java and inheritance. One of my pet hates is string formatting in error messages - it's arbitrary and often doesn't provide enough information. Thus, I've included an extra `data` field to capture the extra data that helps provide more context to error messages. In JVM languages, however, often exceptions are caught by thier super-type (e. `Exception` or `Throwable`). The method `getMessage` is defined on `Throwable` to just return the `message` string and so if I don't serialise the extra map of data, then it'll be missed in any exception logging. I've got to choose some serialisation format so I chose JSON.